ArjunaJTA/jta/tests/classes/com/hp/mwtests/ts/jta/recovery/RecoverySuspendTest.java

```diff
@@ -0,0 +1,205 @@
+    };
+
+    private static RecoveryEnvironmentBean _recoveryConfig;
+    private static JTAEnvironmentBean _jtaEnvironmentBean;
+
+    // fields read with byteman
+    private final static int periodicRecoveryPeriod = 5;
+    private final static int recoveryBackoffPeriod = 1;
+
+    @BeforeAll
+    public static void beforeClass() {
+
+        _recoveryConfig = recoveryPropertyManager.getRecoveryEnvironmentBean();
+        _jtaEnvironmentBean = jtaPropertyManager.getJTAEnvironmentBean();
+
+        // Let's go quick on this
+        _recoveryConfig.setRecoveryBackoffPeriod(recoveryBackoffPeriod);
+        _recoveryConfig.setPeriodicRecoveryPeriod(periodicRecoveryPeriod);
+        // Only one millisecond to stale record as soon as possible
+        _jtaEnvironmentBean.setOrphanSafetyInterval(1);
+        // don't sign off until the store is empty
+        _recoveryConfig.setWaitForWorkLeftToDo(true);
+
+        // the orphan filter
+        XAResourceOrphanFilter xaResourceOrphanFilter = new DummyXAResourceOrphanFilter(XAResourceOrphanFilter.Vote.ROLLBACK);
+        _jtaEnvironmentBean.setXaResourceOrphanFilters(List.of(xaResourceOrphanFilter));
+        // the test set of modules
+        _recoveryConfig.setRecoveryModuleClassNames(Arrays.asList(modules));
+
+        _jtaEnvironmentBean.setXaResourceRecoveryClassNames(
+                List.of(BytemanControlledXAResourceRecovery.class.getName()));
+
+        // obtain a new RecoveryManager with the above config:
+        _manager = RecoveryManager.manager(RecoveryManager.INDIRECT_MANAGEMENT);
+        // recovery can start
+        _manager.startRecoveryManagerThread();
+    }
+
+    @BeforeEach
+    public void beforeEach() {
+        // In case the Recovery Manager was suspended, this will resume it
+        _manager.resume();
+        // Let's clean
+        ThreadActionData.purgeActions();
+    }
+
+    @AfterEach
+    public void afterEach() {
+        if (!TxControl.isEnabled()) {
+            // Re-activate transaction creation
+            TxControl.enable();
+        }
+    }
+
+    @Test
+    @BMScripts(scripts = {
+            @BMScript("RecoverySuspendTest/recoverySuspendTest_BytemanControlledXAResource"),
+            @BMScript("RecoverySuspendTest/recoverySuspendTest_PeriodicRecovery")
+    })
+    public void testSuspensionWhenThereArentTransactionsToRecover() {
+
+        // Make sure that the test environment is ready
+        BytemanControlledXAResource.resetCommitCallCounter();
+        BytemanControlledXAResource.resetRollbackCallCounter();
+        // The following instruction makes the BytemanControlledXAResource commit without failures.
+        BytemanControlledXAResource.setGreenFlag();
+        BytemanControlledXAResource.setCommitReturn(XAResource.XA_OK);
+
+        // BytemanControlledXAResource.getCommitCallCounter() should be 1 as:
+        // - One invocation from the normal commit procedure should pass
+        startTest(1,0, 0);
```

Comment:
I think for this test to be meaningful from the commit point of view, we would want to see the recovery manager re-issue the commit call. Otherwise we can't be sure the XARecoveryModule is even used in this test (hopefully it is [via the AARM] but maybe we can assert that happened)?

Reply:
I think that this test can be removed. WDYT?

Reply:
What do you think about having the test proving the AARM will commit this (as in commit a recovered XAResource)? I think it would be a useful addition.

Reply:
It would be something like: `testSuspensionWhenThereIsAnXAResourceToCommit`

Reply:
What this test does is checking that the new feature of XARecoveryModule isn't blocking the suspension of the Recovery Manager when transactions commit successfully. I proposed to remove this test as I realised that AARM is the actual module that takes care of this logic path...and we already have a test in ArjunaCore. WDYT?

Reply:
I think this is different enough from "pure" AARM to warrant it's own test - at least perhaps. IIRC for the AARM test we did not have an XAR in there. So we don't know for sure that if we have an XAR that needs committing the server will hang waiting for it? Maybe it could through RMFAIL a few times in recovery to make this (new) test prove the server will hang when an XAR needs committing

Reply:
> I think this is different enough from "pure" AARM to warrant it's own test - at least perhaps. IIRC for the AARM test we did not have an XAR in there. So we don't know for sure that if we have an XAR that needs committing the server will hang waiting for it? Maybe it could through RMFAIL a few times in recovery to make this (new) test prove the server will hang when an XAR needs committing

I think I see your point. Thanks Tom.
AFAIU, AARM is the actual RecoveryModule that takes care of XA transactions. While I can definitely develop a new test to check that AARM makes PeriodicRecovery's suspend wait until all XA transactions are completed but it seems to me that we're not actually testing the new functionality of XARecoveryModule. But maybe I'm missing something here :confused:

Reply:
As I said, I see your point. If you prefer that I develop a new test (even though it would be similar to what we already have in AARM), I have no objections.

Reply:
While I believe AARM is the part that will cause the commit of the XAR, I think AARM needs XARM to do it's work and so I don't think it is the only RecoveryModule needing to behave well in order that transactions which have an XAResource can be committed. I am going to look and see if I can be a bit more specific as that might help with the test.

Reply:
OK, so what I think this test would do is:

1. Throw XAER_RMFAIL from the commit which should leave the pending transaction in the object store. This is basically the setup to test what I think needs testing.
2. during suspend, make the first call to XAResource::recover(int) method throw an exception so the suspend has to wait because `recoveryProblems` is a true value (set at https://github.com/jmfinelli/narayana/blob/JBTM-3893/ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/recovery/arjunacore/XARecoveryModule.java#L690). This is to verify that the server wont return from suspend when we have incomplete transactions that will commit.
3. Verify that XAResource::commit will be called a second time (this time by recovery) and this must be before the suspend returns

I don't think the existing AARM test on it's own does this, in particular I doubt it does step 2. You can make an argument that the other test case you are proposing in this file could be made to do this check of L690 in a rollback scenario (so step 2), but a rollback scenario wouldn't do step 3 after doing step 2 and so it feels different to me. Maybe it feels different to me because these two steps somehow speak to confidence in the interactions between AARM and XARM combined with the new RecoveryManager::suspend operation.

Reply:
Thanks for your comment, @tomjenkinson. Hopefully, the new modifications are close enough to the test you described :crossed_fingers: Please, let me know what you think :-)

Reply:
Thank you!